### PR TITLE
Set CompilerApiVersion in Microsoft.Managed.Core.targets

### PR DESCRIPTION
**Pull request by eerhardt**
This allows other targets to understand which compiler will be used.
This is useful in the SDK targets that need figure out the version in order to pick the right analyzer assets.

Contributes to #52265

FYI - @dsplaisted. This resolves the feedback from https://github.com/dotnet/sdk/pull/20793#discussion_r707788803

### CODE REVIEW

File: src/Compilers/Core/MSBuildTask/Microsoft.Build.Tasks.CodeAnalysis.csproj [hunk 1]
```diff
@@ -60,5 +60,45 @@
...
+  <Target Name="GenerateCompilerVersionTargetsFile"
+          Inputs="$(MSBuildThisFileFullPath)"
+          Outputs="$(CompilerVersionTargetsFileIntermediatePath)">
+
+    <PropertyGroup>
+      <CompilerApiVersion>$([System.Version]::Parse($(VersionPrefix)).Major).$([System.Version]::Parse($(VersionPrefix)).Minor)</CompilerApiVersion>
+      <CompilerVersionTargetsFileContent>
+<![CDATA[<Project>
+  <PropertyGroup>
+    <CompilerApiVersion>roslyn$(CompilerApiVersion)</CompilerApiVersion>
```
Comment (eerhardt) [inline src/Compilers/Core/MSBuildTask/Microsoft.Build.Tasks.CodeAnalysis.csproj:92]: NOTE: an alternative, simpler approach could be to straight up hard-code `roslyn4.0` directly into `Microsoft.Managed.Core.targets`. And then rely on the test that I added to break whenever we bump the `Major/MinorVersion` properties in the repo - reminding the person who did the bumping to update the hard-coded version.

I'm open to either approach. I'll let the owners decide which is more preferable (or suggest another alternative).

Comment (chsienki) [reply src/Compilers/Core/MSBuildTask/Microsoft.Build.Tasks.CodeAnalysis.csproj:92]: That's the exact approach I was concerned about, so I'm definitely in favor of doing it this way.

**Review by chsienki — APPROVED** [review]

File: src/Compilers/Core/MSBuildTask/Microsoft.Build.Tasks.CodeAnalysis.csproj [hunk 2]
```diff
@@ -60,5 +60,45 @@
...
+      </Content>
+    </ItemGroup>
+  </Target>
+
+  <Target Name="GenerateCompilerVersionTargetsFile"
+          Inputs="$(MSBuildThisFileFullPath)"
+          Outputs="$(CompilerVersionTargetsFileIntermediatePath)">
+
+    <PropertyGroup>
+      <CompilerApiVersion>$([System.Version]::Parse($(VersionPrefix)).Major).$([System.Version]::Parse($(VersionPrefix)).Minor)</CompilerApiVersion>
```
Comment (jaredpar) [inline src/Compilers/Core/MSBuildTask/Microsoft.Build.Tasks.CodeAnalysis.csproj:88]: Sorry, I realize I should have brought this up sooner but shoudn't we prefix this property name with `Roslyn`? Basically `RoslynCompilerApiVersion` vs. `CompilerApiVersion`? Reasonable for other compilers to provide similar properties.

Comment (eerhardt) [reply src/Compilers/Core/MSBuildTask/Microsoft.Build.Tasks.CodeAnalysis.csproj:88]: > Reasonable for other compilers to provide similar properties.

The reason I didn't put `Roslyn` in property name is exactly for that reason - to support other compilers. The `CompilerApiVersion` version gets piped into the `ResolvePackageAssets` task in the dotnet/sdk:

https://github.com/dotnet/sdk/blob/500e1f61104f94753d7d84c3941d93fec2814291/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets#L267-L272

```
    <ResolvePackageAssets
      ProjectAssetsFile="$(ProjectAssetsFile)"
      ProjectAssetsCacheFile="$(ProjectAssetsCacheFile)"
      ProjectPath="$(MSBuildProjectFullPath)"
      ProjectLanguage="$(Language)"
      CompilerApiVersion="$(CompilerApiVersion)"
```

I didn't want that task to have to understand compiler-specific properties (for example needing to read both `RoslynCompilerApiVersion` or `FSharpCompilerApiVersion`).

Instead, I just made a general `CompilerApiVersion` which can be set to `roslynX.Y` or `fsharpX.Y` or whatever. This way, the task shouldn't need to be updated to support other compilers in the future.

**Pull request merged**